### PR TITLE
feature/Dry-run-as-temporal-activity

### PR DESCRIPTION
**Pull request by lchoquel**
## What

Runs the whole `/validate` job — the validation sweep **and** the graph-producing dry-run — as **one in-process Temporal activity** (`act_dry_validate`) that traces the graph into an **in-memory** event log: one round-trip, library loaded once, no nested Temporal dispatch, no NDJSON files, no DynamoDB tracing I/O, no usage/cost reporting. Direct (non-Temporal) mode stays in-process and unchanged.

This is **Part C** of the dry-run / validation-sweep consolidation (Part A shipped in #956). It is built on the current `temporalio` (1.23.0) — **no SDK bump** — by dispatching a one-step wrapper workflow rather than a standalone activity.

## Why

Under a Temporal-enabled hub, `/validate`'s graph dry-run previously went through `PipelexRunner` → `get_pipe_run()` → a top-level worker workflow that traced to DynamoDB and assembled the `GraphSpec` out of band. That's a full distributed round-trip (plus DDB) for what is conceptually one synchronous, in-process job. Collapsing it into a single activity that runs everything in-process on the worker removes the tracing I/O and the nested dispatch while preserving today's `/validate` semantics.

## How — three additions (see `CHANGELOG.md` `[Unreleased]`)

1. **`scoped_event_log` (`pipelex/hub.py`)** — a ContextVar-scoped, shared trace-event transport (mirrors `scoped_pipe_router`). Both the write side (`pipeline_run_setup`) and the read side (`assemble_tracing`) prefer the scoped instance over building their own via `make_event_log`, so emit + assemble hit **one** `InMemoryEventLog`. This is the fix to the two-instance problem that an in-memory log otherwise can't bridge. A set override implies tracing-enabled; the instance's `close()` must be idempotent/no-op (the write-side tracer closes its transport at teardown). Scoped, so concurrent runs stay isolated.

2. **`dry_run_pipe_in_process` (`pipelex/pipe_run/dry_run_pipeline.py`)** — dry-runs a pipe against an already-open library fully in-process and returns its `GraphSpec`, even under a Temporal hub. Three ContextVar scopes pin the run in-process: `scoped_event_log`, `scoped_pipe_router` (nested controller sub-pipes resolve a local router), and the new `scoped_content_generator` (inference leaves resolve an inline dry generator instead of `ContentGeneratorInWorkflow` — holds whether the DRY mock lives at the pipe or the leaf). Uses the `prepare_pipe_job` seam + a local `PipeRun` — **not** the bridge, **not** `PipelexRunner` (decision D-C1).

3. **`act_dry_validate` + `wf_dry_validate` + `dispatch_dry_validate` (`pipelex/temporal/tprl_pipe/`)** — the activity composes `validate_bundle` (sweep) + the graph entry, loading the library once. The graph half is best-effort: an expected dry-run failure returns `graph_spec=None` with validation still successful. The sweep half is `validate_bundle` itself, so validation failures cross the activity boundary as structured `ErrorReport`s carrying the same categorized `ValidateBundleError` identity the direct path surfaces (a strict-mode signature refusal keeps its offending-pipe + signature refs in the caller-facing message — cross-path parity, decision R-D1). Deterministic failures are marked non-retryable. Registered in the `pipe` task pack.

## Reviewer's guide

- **`TODOS.md`** is the executable plan + progress tracker (cold-start context, file map, per-phase decisions, checkpoint handoffs). Best entry point for understanding the design.
- **`wip/dry-run-refactor/`** — `dry-run-modes-master-plan.md` (big picture), `followup-temporal-validation-activity.md` (this mode's design / why), `followup-dry-validate-hardening.md` (deferred operational hardening).
- Key decisions are inline in `TODOS.md` (D-C1 graph-run shape, R-D1 parity, scope-only in-memory exposure, wrapper-workflow dispatch).

## Verification

- **Mode-1 isolation tests** — `test_dry_validate_activity_in_memory.py` (zero nested dispatch pinned on Temporal *history*, not wrapper spies; validation-failure + no-main-pipe edge cases), `test_dry_run_graph_in_process.py` (zero-dispatch in-memory graph under a Temporal-enabled hub, leaf-level-mock RED-proven), `test_scoped_event_log.py` / `test_scoped_content_generator.py` (ContextVar isolation/nesting/restore), `test_scoped_in_memory_tracing.py` (same-instance emit+read, concurrency), `test_dry_validate_dispatch.py`.
- **Mode-2 Tier 2d** of the `temporal-e2e-validate` skill — verified end-to-end on a real 3-process stack (server + separate worker + submitter), both the activity and API arms **GREEN + RED-proven**; submitter script `scripts/submit_dry_validate.py`.
- `make agent-check` clean; full `make agent-test` green.

## Cross-repo companion (not in this PR)

The `pipelex-api` route split (`_validate_direct` / `_validate_via_temporal`, dispatch-first for error parity) lives on `feature/Update-dry-run-api`. It is **merge-gated on a pipelex release/rev bump > v0.32.1** that ships `act_dry_validate` — its CI pins a released pipelex, so it can't merge until this lands and is released.

## Deferred (tracked, not on this branch's critical path)

- Phase G0: `temporalio` bump → true standalone activity (replacing the one-step wrapper workflow).
- Retire the old worker-workflow + DynamoDB graph path for `/validate` once the API cuts over.
- Operational hardening items in `wip/dry-run-refactor/followup-dry-validate-hardening.md`.
- Mode C (full-distribution leaf mocks) is a separate, unbuilt mode; the `scoped_content_generator` seam it needs now exists.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Runs the whole `/validate` job (sweep + graph dry‑run) as one in‑process Temporal activity and unifies dry‑run across backends. One round‑trip, no nested dispatch, and in‑memory tracing for a consistent graph regardless of host tracing config.

- New Features
  - `act_dry_validate` with wrapper workflow `wf_dry_validate` returns `{status map, GraphSpec | None}` in one call; the library loads once; best‑effort graph; deterministic validation errors are non‑retryable and the workflow sets `RetryPolicy(maximum_attempts=1)`.
  - Unified dry‑run: `run_mode` rides a derived `CogtRunParams` stamped on every leaf; operators reuse the live path; `ContentGeneratorDry` removed. DRY branches at LLM, extract, img‑gen, search, render, and templating leaves; a Temporal DRY run dispatches real `act_*` and mocks inside them. Hidden DRY sub‑flag `is_mock_usage` (CLI `--mock-usage`) emits synthetic usage for cost‑report tests.
  - `dry_run_pipeline` owns its graph transport via `scoped_event_log`; added `DryRunGraphNotProducedError` when a graph is expected but missing. Validation sweep and in‑process dry‑runs pin an inline generator via `scoped_content_generator`; submitter `dispatch_dry_validate` sets explicit timeouts.

- Bug Fixes
  - Hardened tracer setup: `GraphTracerManager.open_tracer` registers the tracer only after `setup()` succeeds to avoid leaking unusable entries; tests updated to use `scoped_current_library()`. `scripts/submit_dry_validate.py` now expands `~` in `--bundle` / `--graph-out` paths.
  - `scripts/submit_dry_validate.py`: GraphSpec JSON write is best‑effort — prints stats first, creates parent dirs, and degrades gracefully on write errors without changing exit codes.
  - Clarified the hidden `--mock-usage` help text across `pipelex run` commands to avoid ambiguity.

<sup>Written for commit 769b6b1b3e4781b31a415bb4cd90e0e15919dfeb. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Pipelex/pipelex/pull/982?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: pipelex/pipe_run/dry_run_pipeline.py [hunk 1]
```diff
@@ -84,3 +95,92 @@ async def dry_run_pipeline(
...
+    finally:
+        # Safety net for failures before PipeRun.run owns the tracer (e.g. prepare_pipe_job
+        # raising): close_tracer pops by key, so this is a no-op after a completed run.
+        graph_tracer_manager.close_tracer(pipeline_run_id)
+
+    if not pipe_output.graph_spec:
+        msg = f"In-process dry-run of pipe '{pipe.pipe_ref}' did not produce a graph spec"
+        raise PipelexError(msg)
+
+    return pipe_output.graph_spec
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply pipelex/pipe_run/dry_run_pipeline.py:191]: ➖ False positive — the `try/finally` has no `except` clause, so an exception raised by `prepare_pipe_job()` propagates out of `dry_run_pipe_in_process` after the `finally` closes the tracer. The post-`try` read of `pipe_output` (line 182) is only reachable on the happy path, where `pipe_output` is always bound — no `UnboundLocalError` is possible. The exception reaches `act_dry_validate`'s catch for `(PipelexError, ValidationError, FactoryException)` exactly as designed, degrading to `graph_spec=None`. The suggested `pipe_output = None` initialization would also make the `is None` check dead code, and our coding standards deliberately keep multi-path variables unbound so linters catch uncovered paths.

File: wip/dry-run-refactor/dry-run-modes-master-plan.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+- **Mode 2 wants that default** — the in-workflow generator *is* how it dispatches leaf activities.
+- **Mode 1 must override it to inline** — otherwise, once Part B routes the DRY mock through `get_content_generator()` at the leaf, Mode 1's in-process activity would dispatch `act_llm_gen_*` and break.
+
+So `scoped_content_generator` (a hub ContextVar + context manager, mirroring `scoped_pipe_router` at `hub.py:635`, with `get_content_generator()` preferring the override) is needed by Mode 1 and is the seam Part C originally specced. **Decision (2026-06-09): build it now, in Mode 1's branch**, so Mode 1 is correct regardless of when Part B lands and the seam exists once for both.
+
+### The ContextVar scope family (`pipelex/hub.py`)
+
+| Scope | Exists? | Used by | Forces |
+|---|---|---|---|
+| `scoped_current_library` | yes (`hub.py:518`) | both / general | which library is current |
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply wip/dry-run-refactor/dry-run-modes-master-plan.md]: ✅ Fixed in 6763a7ea7 — confirmed both refs were stale (`scoped_current_library` is at hub.py:546, `scoped_pipe_router` at hub.py:663). Rather than re-pinning numbers that will drift again, dropped the line numbers and kept the greppable function names (also fixed the same stale `hub.py:635` in the prose above the table).

File: tests/integration/pipelex/temporal/test_dry_run_graph_in_process.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    async def _run_in_process_graph_dry_run(self, library_id: str) -> GraphSpec:
+        mthds_content = Path(PipeParallelTemporalTestData.BUNDLE_FILE).read_text(encoding="utf-8")
+        acquire_library(library_id=library_id, mthds_contents=[mthds_content])
+        try:
+            main_pipe = get_required_pipe(pipe_code=_PARALLEL_MAIN_PIPE_REF)
+            return await dry_run_pipe_in_process(pipe=main_pipe, library_id=library_id)
+        finally:
+            get_library_manager().teardown(library_id=library_id)
+            clear_current_library()
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply tests/integration/pipelex/temporal/test_dry_run_graph_in_process.py]: ✅ Fixed — the helper now wraps the run in `scoped_current_library()` so any outer current-library binding is restored on exit instead of being unconditionally cleared.

File: wip/dry-run-refactor/dry-run-modes-master-plan.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+So `scoped_content_generator` (a hub ContextVar + context manager, mirroring `scoped_pipe_router` in `hub.py`, with `get_content_generator()` preferring the override) is needed by Mode 1 and is the seam Part C originally specced. **Decision (2026-06-09): build it now, in Mode 1's branch**, so Mode 1 is correct regardless of when Part B lands and the seam exists once for both.
+
+### The ContextVar scope family (`pipelex/hub.py`)
+
+| Scope | Exists? | Used by | Forces |
+|---|---|---|---|
+| `scoped_current_library` | yes (`hub.py`) | both / general | which library is current |
+| `scoped_pipe_router` | yes (`hub.py`, PR #976) | Mode 1 | nested controllers run in-process, not via the Temporal router |
+| `scoped_content_generator` | **to build (Mode 1 branch)** | Mode 1 | the leaf uses the inline generator, not the in-workflow one |
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply wip/dry-run-refactor/dry-run-modes-master-plan.md]: ✅ Fixed — the scope-family table now marks `scoped_content_generator` / `scoped_event_log` as built, and the build-order section reflects Phases 1–3 as done (Phase 4 cross-repo dispatch stays pending).

File: .claude/skills/temporal-e2e-validate/scripts/submit_dry_validate.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    print(f"STATUS MAP ({len(result.dry_run_outputs)} pipes):")
+    for pipe_ref, output in sorted(result.dry_run_outputs.items()):
+        suffix = f" — {output.error_message}" if output.error_message else ""
+        print(f"  {output.status:<8} {pipe_ref}{suffix}")
+
+    if result.graph_spec is None:
+        print("GRAPH: None (best-effort — validation still succeeded)")
+    else:
+        graph_path = Path(graph_out)
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply .claude/skills/temporal-e2e-validate/scripts/submit_dry_validate.py]: ✅ Fixed — `--graph-out` is now `expanduser()`d before writing.

File: .claude/skills/temporal-e2e-validate/scripts/submit_dry_validate.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+from pipelex.temporal.tprl_pipe.act_dry_validate import DryValidateArg
+from pipelex.temporal.tprl_pipe.dry_validate_dispatch import dispatch_dry_validate
+
+DEFAULT_BUNDLE = "tests/integration/pipelex/temporal/library_crate/temporal_parallel.mthds"
+DEFAULT_PIPE = "temporal_parallel_test.temporal_parallel_sequence"
+DEFAULT_GRAPH_OUT = "/tmp/tier2d-graph-spec.json"  # noqa: S108
+
+
+async def submit(bundle: str, pipe_code: str | None, allow_signatures: bool, graph_out: str) -> int:
+    bundle_content = Path(bundle).read_text(encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply .claude/skills/temporal-e2e-validate/scripts/submit_dry_validate.py]: ✅ Fixed — `--bundle` is now `expanduser()`d before reading.

File: pipelex/pipe_run/dry_run_pipeline.py [hunk 7]
```diff
@@ -84,3 +95,92 @@ async def dry_run_pipeline(
...
+    # workflow). Keep the router instance: the scope installs it for nested sub-pipes.
+    # Constructed BEFORE the tracer opens so nothing can raise between open_tracer and the
+    # try/finally that guarantees its close — an entry in the process-wide GraphTracerManager
+    # must never outlive this call.
+    pipe_router = PipeRouter(observer=ObserverNoOp())
+    pipe_run = PipeRun(pipe_router=pipe_router)
+    content_generator = ContentGeneratorDry()
+
+    graph_tracer_manager = GraphTracerManager.get_or_create_instance()
+    trace_context = graph_tracer_manager.open_tracer(
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply pipelex/pipe_run/dry_run_pipeline.py:157]: ✅ Fixed at the root cause — `GraphTracerManager.open_tracer` now registers the tracer only after `setup()` fully succeeds, so a setup-time failure can never leak an entry for any caller (not just this one). Regression unit test added in `tests/unit/pipelex/graph/test_graph_tracer_manager.py`.

File: wip/dry-run-refactor/dry-run-modes-master-plan.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (lchoquel) [reply wip/dry-run-refactor/dry-run-modes-master-plan.md]: ✅ Fixed — duplicate of the other doc-status thread; the plan doc now reflects implemented status.

File: pipelex/cogt/content_generation/assignment_models.py [hunk 8]
```diff
@@ -18,26 +19,30 @@
 
 class LLMAssignment(BaseModel):
     job_metadata: JobMetadata
+    cogt_run_params: CogtRunParams
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply pipelex/cogt/content_generation/assignment_models.py:22]: ➖ False positive — the Temporal integration has never shipped to production, so there are no in-flight payloads to deserialize against the new schema. This project also explicitly does not maintain wire backward compatibility (workers and submitters are deployed in lockstep from the same pinned version).

File: pipelex/cli/commands/run/bundle_cmd.py [hunk 9]
```diff
@@ -61,11 +61,12 @@ def run_bundle_cmd(
...
     ] = False,
-    mock_inference: Annotated[
+    mock_usage: Annotated[
         bool,
         typer.Option(
-            "--mock-inference",
-            help="Live run that fakes AI calls at the inference leaf with reportable synthetic usage. Mutually exclusive with --dry-run.",
+            "--mock-usage",
+            hidden=True,
+            help="Internal test trigger: dry run whose LLM leaves report non-zero synthetic usage so the cost report renders. Requires --dry-run.",
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply pipelex/cli/commands/run/bundle_cmd.py]: ✅ Fixed in 68832220b — the sentence was grammatically valid ("leaves" is the plural of leaf, "report" is the verb) but a real garden-path; reworded to "dry run whose LLM leaf mocks report nonzero synthetic usage" in all three run commands (bundle, pipe, method).
